src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs

```diff
@@ -91,26 +101,63 @@ private static void ClearCachedData()
 
-            await SearchCachedDocumentsInCurrentProcessAsync(
-                storageService, patternName, patternContainer, declaredSymbolInfoKindsSet,
-                onItemFound, priorityDocumentKeys, cancellationToken).ConfigureAwait(false);
+            // Process the documents by project group.  That way, when each project is done, we can
+            // report that back to the host for progress.
+            var groups = documentKeys.GroupBy(d => d.Project).ToImmutableArray();
 
-            await SearchCachedDocumentsInCurrentProcessAsync(
-                storageService, patternName, patternContainer, declaredSymbolInfoKindsSet,
-                onItemFound, lowPriDocs, cancellationToken).ConfigureAwait(false);
+            using var _1 = GetPooledHashSet(priorityDocumentKeys, out var priorityDocumentKeysSet);
+
+            // Sort the groups into a high pri group (projects that contain a high-pri doc), and low pri groups (those
+            // that don't).
+            using var _2 = GetPooledHashSet(groups.Where(g => g.Any(priorityDocumentKeysSet.Contains)), out var highPriorityGroups);
+            using var _3 = GetPooledHashSet(groups.Where(g => !highPriorityGroups.Contains(g)), out var lowPriorityGroups);
+
+            await ProcessProjectGroupsAsync(highPriorityGroups).ConfigureAwait(false);
+            await ProcessProjectGroupsAsync(lowPriorityGroups).ConfigureAwait(false);
+
+            return;
+
+            async Task ProcessProjectGroupsAsync(HashSet<IGrouping<ProjectKey, DocumentKey>> groups)
+            {
+                using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
+
+                foreach (var group in groups)
+                    tasks.Add(ProcessProjectGroupAsync(group));
+
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+            }
+
+            async Task ProcessProjectGroupAsync(IGrouping<ProjectKey, DocumentKey> group)
+            {
+                await Task.Yield();
+                var project = group.Key;
+
+                // Break the project into high-pri docs and low pri docs.
+                using var _1 = GetPooledHashSet(group.Where(priorityDocumentKeysSet.Contains), out var highPriDocs);
+                using var _2 = GetPooledHashSet(group.Where(d => !highPriDocs.Contains(d)), out var lowPriDocs);
+
+                await SearchCachedDocumentsInCurrentProcessAsync(
+                    storageService, patternName, patternContainer, declaredSymbolInfoKindsSet,
+                    onItemFound, highPriDocs, cancellationToken).ConfigureAwait(false);
+
+                await SearchCachedDocumentsInCurrentProcessAsync(
+                    storageService, patternName, patternContainer, declaredSymbolInfoKindsSet,
+                    onItemFound, lowPriDocs, cancellationToken).ConfigureAwait(false);
+
+                // done with project.  Let the host know.
+                await onProjectCompleted().ConfigureAwait(false);
```

Comment:
>onProjectCompleted [](http://example.com/codeflow?start=22&length=18)

So I'm clear, this gets fired groups.Count x 2 times?

Reply:
discussed offline.  thsi should be called groups.Count times.